internal/worker/client.go

```diff
@@ -121,9 +121,17 @@ func NewClient(conf ClientConfig) (*Client, error) {
 		clientId:     conf.ClientId,
 		clientSecret: conf.ClientSecret,
 	}
-	err = client.registerWorker()
+
+	// get a token first to avoid an "unauthorized" reply
+	// essentially saving one request on startup
+	err = client.refreshAccessToken()
 	if err != nil {
-		return client, err
+		logrus.Warnf("Error getting access token on startup, %v. Trying again later…", err)
+	} else {
+		err = client.registerWorker()
+		if err != nil {
+			logrus.Warnf("Error registering worker on startup, %v. Trying again later…", err)
+		}
```

Comment:
There's a lot more stuff here than before now. I think it's fine to let the lazy token do it's thing, a single 401 request at the start is really cheap. Getting the worker heartbeat to register immediately is tricky? Maybe just move the registerWorker call into the `workerheart` fn right before the loop?

Reply:
for the tests this is not really deterministic I guess - otherwise this could work

Reply:
I thought it's cleaner to do the token and registration in sequence, in the normal "good case" and retry and do lazy otherwise

Reply:
Hmm we do lazy refreshes basically everywhere. And if it registers at the very start it should be pretty deterministic no? Like 1 call with a 401, and 1 successful? The accesstoken is behind a mutex so as long as the client does one operation at a time.

Reply:
I mean doing it before the loop in `workerHeartbeat` might be non-deterministic with the main-loop who ever gets dispatched first (e.g. during tests)

Reply:
It the possible option:

- do "registerWorker()" as before and instead of returning an error just log a warning

covered above? That one seems fine, it's a small change, it does stop the startup crash. The call will register and as part of the lazy refresh a token. if that fails the heartbeat will try again in a minute. or am I missing something?

Reply:
yea I'd lean towards that one as well, keeps the `New*()` simple and does the lazy token refresh as we do elsewhere.

Reply:
I think there are three options
* explicitly get a token on startup, register, do lazy/token recovery later (current implementation of this PR)
   no 401 expected - only in case of network/deployment issues (during startup or later)
   also the behavior in case we can't contact composer is the same on startup or during runtime (just retry)
* get a token before going into workerHeartbeat-loop
   possibly racy if the client will do this first or the main loop will try a request first and lazy get a token
   here the startup might generate 401 or not
* just start the heartbeat (acting in 1min) and going to main-loop
   creates a 401 first for sure, which could be ok and works,
   but not necessary to generate an "error" in the "happy path" IMHO

Did I get them right?

regarding the behavior in general and related tests I think the first option is the cleanest and most predictable startup

Reply:
I thought about this a little bit and one nice aspect of doing the lazy refresh is that we ensure it works when doing it this way. It may sound silly but let's think about it: if we somehow break lazy refresh of the token we currently get errors left and right at startup because nothing will work and we will know immediately. If however we get the token explicitly and lazy-refresh-on-401 is broken somehow we will notice that only after the 900s time that suddenly something start failing and there are probably not that many things doing on after 900s so it maybe way later. So the pattern to spot what went wrong is harder to find.

Reply:
(Ok - out of context - but still my point 😇)
> we currently get errors left and right at startup

But that's exactly my point. In the current implementation we _always_ get a 401 (incl. an warning which you assert in your test) even in the happy path.

Otherwise we only get a 401 and a refresh in case of connection errors, that seems more straight forward to me.